Clean up handling of Icons under unified LabIcon

telamonian opened the pull request:
## References

Fixes #7765

As part of #7700, I added a `@jupyterlab/ui-components` dependency to `@jupyterlab/rendermime-interfaces`. This was done in order to use `JLIcon` as the type of `IFileType.iconRenderer`. However, that brought in a number of unwanted sub-dependencies to rendermime-interfaces, such as React. This PR restores the isolation of rendermime-interfaces by:

- Adding a minimal `IJLIcon` interface definition
- Hard copying the `IJLIcon` definition over to rendermime-interfaces (as was done with `IFileType`, etc)

## Code changes

In `JLIcon`, I removed the `class` method and refactored `recycle` into a static method called `remove`. Both methods were effectively already static methods, and `class` was redundant with the `iconStyle` function. These tweaks are needed for this PR, since they also fix some issues due to it no longer being safe to assume that `IFileType.iconRenderer` has the full functionality of `JLIcon`. 

## User-facing changes

None

## Backwards-incompatible changes

The changes to the JLIcon methods are backwards incompatible. Since they're fixes, and since we're still in beta, I think they should go in

jasongrout commented on packages/rendermime-interfaces/src/index.ts:
```diff
@@ -227,6 +228,20 @@ export namespace IRenderMime {
+    /**
+     * Create an icon as a DOM element
+     *
+     * @returns A DOM element that contains an (inline) svg element
```
Does it *have* to be an inline svg element? Or might that change in the future, and that is just an implementation detail? If it is just an implementation detail, it should not be in the interface documentation.

This is an interface we do not want to break, because we want to basically never have a backwards incompatible change in rendermime-interfaces. So we should be as generic as we can here and still obtain our objective.

telamonian replied:
> because we want to basically never have a backwards incompatible change in rendermime-interfaces

Ah. I was not aware of that detail. In that case, we should probably do away with the `IJLIcon.element` function entirely. Basically, `element` is in that interface in order to account for the fact that the file type icons are handled differently in the tabbars and in the filebrowser listing. If I fix things up so that `render` and `unrender` can be used to create the icons in both contexts, that eliminates the need for `element` here.

telamonian replied:
I've now removed `IJLIcon.element`

blink1073 commented on packages/rendermime-interfaces/src/index.ts:
```diff
@@ -227,6 +228,20 @@ export namespace IRenderMime {
 
+  /**
+   * The IJLIcon interface, which supplies element, render,
+   * and unrender functions
```
I think we should mention here that there is an implementation and some example renderers in `@jupyterlab/ui-components`.

jasongrout replied:
Thanks! I was hoping to see an example of how it is used.

jasongrout commented:
Would it be possible to not have a dependency on a virtual dom element? Can we abstract out that detail, and make it so they can just supply the svg icon?

telamonian commented:
> Can we abstract out that detail, and make it so they can just supply the svg icon?

It depends on what exactly you mean by "the svg icon". If you mean "a string with the svg src", they can already more-or-less do that:

```typescript
iconRenderer: new JLIcon({ name, svgstr });
```

We could also add some sugar to also allow for:

```typescript
iconRenderer: { name, svgstr };
```

but I can forsee some problems that may cause. I'd prefer to encourage users to pre-define their icons (as JLIcon instances) and then pass one of those to `iconRenderer`.

Is that along the lines of what you thinking, or did you mean something else?

jasongrout commented on packages/extensionmanager/src/widget.tsx:
```diff
@@ -19,6 +19,7 @@ import ReactPaginate from 'react-paginate';
 
 import { ListModel, IEntry, Action } from './model';
 import { isJupyterOrg } from './query';
+import { iconStyle } from '@jupyterlab/ui-components/lib/style/icon';
```
We should never import from `lib`. Something like the above sometimes happens to me when the typescript autocomplete fills out the import, and I have to change it back to something like`import { iconStyle } from '@jupyterlab/ui-components';`

telamonian replied:
Fixed! Turns out I didn't have the exports from the `src/style` subpkg set up correctly in the various `index.ts` files, which is why the autocomplete went for the funky `.../lib/...` import in the first place

jasongrout commented:
> Is that along the lines of what you thinking, or did you mean something else?

What I meant was that the concept of a virtual dom is quite advanced for many users, and it looks like we are forcing them to think about virtual dom elements in order to have icons. Can we make the interface simpler by *just* supporting something like `{name, svgstr}`? If they have to create a `JLIcon` to easily use this icon functionality, then that means their plugin now depends on ui-components, which is essentially the same as rendermime-interfaces depending on ui-components, so we haven't gained anything by removing the dependency, practically speaking.

Just curious, what is the benefit of using `new JLIcon({ name, svgstr });` over `{ name, svgstr };`?

jasongrout commented:
I looked at the ui-components package a bit more. To rephrase my question/concerns above: what would be the disadvantage of deleting the `@lumino/virtualdom` dependency, and making rendermime-interface icon *just* accept a `{name, svgstr}` pair? The advantage is that it is very simple to understand (no knowledge of virtual dom needed) and we cut out one more dependency (and apparently we eliminate depending on ui-components for many people too, which is an even bigger issue).

jasongrout commented:
In fact, I'm not even sure the `name` field is needed if we aren't reusing these icons. How about we just have:

```

    /**
     * The icon class name for the file type.
     */
    readonly iconClass?: string;

    /**
     * The icon label for the file type.
     */
    readonly iconLabel?: string;

    /**
     * The svg string for the icon.
     */
    readonly icon?: string;
```

telamonian commented:
> then that means their plugin now depends on ui-components, which is essentially the same as rendermime-interfaces depending on ui-components

I think it would help the discussion if I understood the use case you have in mind a bit better. Can you point me to some examples of plugins that do depend on rendermime-interfaces but don't depend on eg application (which in turn depends on ui-components)?

Also, I was wondering: given the desired level of isolation of rendermime-interfaces w.r.t. the rest of the `@jupyterlab` monorepo, why is it not in `@lumino` instead? Though this may be a discussion for another issue.

telamonian commented:
> What I meant was that the concept of a virtual dom is quite advanced for many users, and it looks like we are forcing them to think about virtual dom elements in order to have icons.

`IRenderer` happens to live in the virtualdom package, but it doesn't really have anything to do directly with virtual DOM:

```
export type IRenderer = {
    render: (host: HTMLElement) => void,
    unrender: (host: HTMLElement) => void
  };
```

No advanced knowledge is required, and this way users can easily set up svgs using the old as-css-background pattern, if they so desire (though I can see an argument for making `unrender` optional):

```
const fooIcon = {
    render: (host: HTMLElement) => host.className = 'my-foo-icon',
    unrender: (host: HTMLElement) => void
  };
```

Really though, I'm hoping to encourage that most people just use `JLIcon`. On that track, from your perspective would it help if JLIcon were split up so as to have a non-react using base/sibling class? Because then `JLIcon` could be used in rendermime-interfaces without adding any major deps.

jasongrout commented:
> I think it would help the discussion if I understood the use case you have in mind a bit better. Can you point me to some examples of plugins that do depend on rendermime-interfaces but don't depend on eg application (which in turn depends on ui-components)?

All of the renderers in https://github.com/jupyterlab/jupyter-renderers/tree/master/packages except for the mathjax and katex renderer. The pdf renderer in core. I think the main intended usecase for rendermime-interface was renderers that don't depend on anything else from JupyterLab.


> Also, I was wondering: given the desired level of isolation of rendermime-interfaces w.r.t. the rest of the `@jupyterlab` monorepo, why is it not in `@lumino` instead? Though this may be a discussion for another issue.

We could possibly do that, but it seems to fit more naturally into JLab, since it's a very restricted stable interface for a JLab plugin for rendering mimetypes in JLab. I mean, it is pretty specific with the notion of filetypes, etc., which fit more with JLab than more general lumino, I think.

jasongrout commented:
> Really though, I'm hoping to encourage that most people just use `JLIcon`.

The easiest way to do that, I think, is to just make the icon an svg string, and *we* wrap it in a JLIcon.

As for using css - we already have an iconClass field too - they could just use that, right?

Again, what would be the disadvantage of just having `icon` be an svg string (and we create the JLIcon when we create the rendermime plugin from their extension)? I don't see a disadvantage, so I'm not sure what we would be trading off doing this.

jasongrout commented:
> We could also add some sugar to also allow for:
> 
> ```ts
> iconRenderer: { name, svgstr };
> ```
> 
> but I can forsee some problems that may cause.

What problems did you foresee?

telamonian commented:
@jasongrout I'm prepping an answer for you. Part of the reason why it's tricky to think through is that I haven't written anything down about the design of JLIcon (partly because it's been in flux until recently). To that end I'm also adding some some docs about JLIcon (for now, I'll stick them in the ui-components README).

Would you happen to have some time to talk about icons tomorrow? I think it would be good for us to sync up about this, and it'll probably be difficult for me to explain all of it during the weekly meeting.

jasongrout commented:
Thanks! I can't talk tomorrow (and won't make the meeting tomorrow), but could talk on Thursday.

telamonian commented:
> As for using css - we already have an iconClass field too - they could just use that, right?

I plan to deprecate the `iconClass` field in `IFileType` and all similar interfaces, with the goal of removing it in jlab 3.0 (I kind of want to see how the JLIcon rollout goes first, though). The current uses of `iconClass` (to set a background image and/or styling for said image via css class) are redundant w.r.t. the functionality of JLIcon. Getting rid of `iconClass` reduces the possibility of weird styling clashes, would simplify some implementation issues related to dynamic icon lookup, and would be one more step towards the goal of having exactly one simple (as possible) way to set up and use icons in jlab.

Originally, I was just going to remove `iconClass` as part of implementing JLIcon. I hesitated when I realized that doing so would make the jlab 2.0 migration process much rockier for many existing jlab extensions (ie any extension that passes an icon into core, so anything with a filetype, settings, sidebar, toolbar button, etc icon). In the end, I implemented a slightly refined version of the backwards compatibility code brought up in #7685:

- For an icon named `foo`, the icon lookup code will match to both `foo` and `jp-FooIcon`
- When `iconRenderer` is not specified, dynamic lookup is performed by looping through all class names in `iconClass` and returning the first match 
- If there are no matches, create a blank element with class `iconClass` (ie use the old icon-as-background-image behavior)

jasongrout commented:
Thanks, and thanks for writing things up in #7782. I'm convinced even more that the rendermime interfaces should not have any notion of jlicon after reading through the above comments and your very helpful writeup:

1. The JLIcon interface is still iterating and likely to change. If a rendermime interface needs to create a JLIcon, it will need to be updated when there are changes to how JLIcons work.
2. It's not likely that a rendermime extension will need to define an icon that someone else in the system will need to use independently
3. Anyways, a major point here is that the rendermime interface is very simple and constrained. An svg string fills that goal.

I'm still curious: What exactly is the benefit of a rendermime extension creating a JLIcon, as opposed to giving an svg string and us creating the JLIcon for it?

jasongrout commented:
@telamonian - it seems that you have reverted the rendermime-interface changes around svg icons completely? Are we going back to just the icon class, or do we still have a way to create an svg icon from a rendermime interface?

telamonian commented:
@jasongrout I don't want this corner case to block the release of the 2.0 rc. Also, rendermime-interfaces is supposed to be a "forever" api. Thus, I think the best thing to do for right now is to not rush and just revert rendermime-interfaces to its pre-#7700 state. The current version of this PR will do that (with partial support for svg icons via matching to the `IFileType.iconClass` param).

As you said, my thinking on at least certain aspects of the design of JLIcon continues to change. Based on the issues and alternatives that you've brought up, here are two possible icon interfaces for rendermime-interfaces that both meet your want of a way to pass in a simple svg string and (and probably most of) my concerns about functionality:

- two param:
    ```
    icon?: {name: string, svgstr: string};
    iconName?: string;
    ```
    This has the advantage of clearly separating the intent to define a new icon (by setting `icon`) from the intent to reuse an existing icon (by setting the `iconName`).

- one param:
    ```
    icon?: string | {name: string, svgstr: string};
    ```
    This has the advantage of more accurately conveying the idea of icon as a one-of: it is either a string with the icon name (to reuse an existing icon), or it is an object with a `{name, svgstr}` pair (to define a new icon). Disadvantage: it leaves somewhat ambiguous whether the bare `string` argument is meant to be the icon name (yes) or the icon's raw svg str (no).

I'd like to develop at least one of these ideas further in a followup PR targeting 2.1. Also at that point we'd deprecate `iconClass`.

telamonian commented:
@jasongrout @blink1073 @saulshanabrook 

I've removed the `iconClass` backwards compatibility cludge (at least w.r.t. filetype icons), and I've added a working implementation of the:

```typescript
icon: string | {name: string, svgstr: string}
```

interface to both versions of the `IFileType` interface (ie in both the docregistry and rendermime-interfaces pkgs).

The PR still needs a lot of work, but I'm at the point where some feedback would be helpful (and would also probably help me finish this up faster). Here are my primary concerns from a design perspective:

- My current strategy for handling the `icon` in file type interfaces goes like so:
    - When a file type is registered, `DocumentRegistry` will now create and work with a `FileType` object, instead of just storing and working with the initial `IFileType` directly.
    - As part of the `FileType` constructor, `IFileType.icon` is "upconverted" into a fully fledged `JLIcon`. The conversions are handled by a pair of static methods, `FileType.resolve` and `JLIcon.resolve`. 
    - Here's the problem. The various document extensions interact with file types using many variations on the basic pattern. I ended up having to call `FileType.resolve` in many different pkgs, basically in order to paper over situations in which I may have been given an `IFileType` but I actually need a `FileType`. The frequent usage of `FileType.resolve` has a bit of a smell. Is there a more elegant way that I can be handling these cases overall?

- I added a way for advanced users to set their own `render` and `unrender` methods when creating a new `JLIcon` (via some extra optional fields).
    - I'm also working on adding those fields into the `IFileType.icon` interface, to allow for advanced customization of icon behavior without needing to interact directly with `JLIcon` at all. This should provide a way to cover many of the edge cases that were brought up in the meeting last Wed (eg render bitmap icons by setting a png as a css background image). 
    - It's currently a bit of a mess, and I have the feeling there may be a better way to set that up compared to what I currently have in place. Another pair of eyes would be welcome.

Any input on the above two concerns, and also just a general review of the code so far, would be greatly appreciated.

telamonian commented:
Here's what's left to do (I'll add to this post as it occurs to me):

- Several more cleanup passes of the code
- Finish adding optional `render` and `unrender` fields to the `IFileType.icon` interface

telamonian commented:
Oh, and also I added what should be a fully functional implementation of dynamic svg replacement/re-rendering. It triggers whenever you set the `.svgstr` param of an `JLIcon` instance. Still needs a lot of testing, though. I have some ideas about adding a `test/ui-components` pkg.

blink1073 commented:
I think the simplified `.icon` should only exist on `rendermime-interfaces`.  Otherwise you should always be using an `.iconRenderer` that is able to be given directly to `title.iconRenderer`.  It is too easy to forget to upconvert otherwise.
I agree that you shouldn't always need a `LabIcon`, just something that meets the `IRenderer` interface.

telamonian commented:
> I think the simplified `.icon` should only exist on `rendermime-interfaces`

@blink1073 Good call. That let me simplify the implementation considerably, and completely revert all uses of `FileType` back to just `IFileType`. Upconversion of icons now happens automagically only in one place, during the creation routine for rendermime plugins.

telamonian commented:
There's now some fixes to `@lumino/virtualdom` that are related to this PR up at jupyterlab/lumino#44. The relevant part is that, if a virtual node has a custom renderer set, an `options` object will now be passed to that renderer's `.render` method whenever it gets called. This PR should be forwards compatible with those changes in Lumino since it adds an `options` param to the corresponding interface in `JLIcon`, ie `IRenderer.render`.

saulshanabrook commented on packages/ui-components/src/icon/jlicon.tsx:
```diff
@@ -333,26 +460,128 @@ export class JLIcon {
    */
-  export interface IOptions {
-    name: string;
+  export interface IIcon {
```
Do we need this interface since we already have one in rendermine? Can we import that one here?

telamonian replied:
Originally I was just following the same convention as for IFileType: the "real" interface lives in docregistry, while rendermime-interfaces contains a "seperate but equal" reimplmentation. On reflection, "because IFileType did it" doesn't make for the strongest design argument. @jasongrout What's your opinion on how interface separation should work here?

telamonian replied:
I do kind of like having the two separate `IIcon` interfaces:

- rendermime-interfaces obviously can't depend on ui-components, but it doesn't really make that much sense for ui-components to depend on rendermime-interfaces

- This way, I'm free to make tweaks to the "real" `IIcon` while the rendermime-interfaces `IIcon` remains forever frozen in amber (of course, I could acheive more or less the same thing by importing `IIcon` from rendermime and then extending it, but for such a simple interface I think it's kind of a wash).

blink1073 replied:
I think in general we should prefer to inherit from `rendermime-interfaces` to keep us honest.

blink1073 commented on tests/test-ui-components/src/labicon.spec.ts:
```diff
@@ -0,0 +1,14 @@
+
+import { LabIcon } from '../..//ui-components/icon';
+
+describe('@jupyterlab/nbformat', () => {
```
`ui-components`*

telamonian replied:
Thanks! That was pretty messed up. Fixed

telamonian commented:
This should now be good to go, and is ready to be pulled in.

There's still a couple of warts (in the form of 2 `UNSTABLE_` methods that are needed to cover exactly 3 corner cases), and a bunch more tests I'd like to write, but:

- `IFileType.icon: string | {name, svgstr}` in rendermime-interfaces is now fully functional
- the `iconClass` => icon name automagic conversion shim is now completely removed (aside from in one of the aforementioned special cases, since I can't easily add an `icon` field to cmds from the command palette)
- everything works!

so there's no reason this should continue to block the RC.

telamonian commented:
Also, there's now some actual documentation! There's user docs for `LabIcon` in the README, while CONTRIBUTING has some dev reference about `LabIcon` development (mostly they're notes for my own benefit, but may be of interest to others)

saulshanabrook commented:
@telamonian nice guide! It seems like the stuff in the README could be moved to https://jupyterlab.readthedocs.io/en/stable/developer/extension_points.html so other extension authors can easily reference it?

telamonian commented:
@blink1073 I've updated the docs. There's now info in the extension point docs, and a note/link in the extension migrations docs.

telamonian commented:
>  I fixed the docs

Ah. I could not for the life of me figure out the right type for typescript + react code in rst. Thanks Steve!

blink1073 commented:
@telamonian, are you looking into the virtualdom signature issue mentioned on [gitter](https://gitter.im/jupyterlab/jupyterlab?at=5e31c58fd9895b17c3af244d)?

telamonian commented:
@blink1073 I am, and I'm working on it right now. The short version is that this PR is not directly related to the problem, but it currently doesn't fix it either. I think I have a fix, just need to test it out

telamonian commented:
@blink1073 I am, and I'm working on it right now

telamonian commented:
@blink1073 I have it worked out. I'll try to upload the fix early tomorrow

blink1073 commented:
👍

blink1073 commented:
@telamonian, were there no changes required in `lumino`?

telamonian commented:
> were there no changes required in `lumino`?

Nope, no changes to Lumino were needed. If we pull this PR in and then do a new release that will fix all of the typing issues reported on gitter (though extension devs will need to update their `package.json` to point to said new release).

There's a separate issue as to whether the recent changes to `@lumino/virtualdom` (in jupyterlab/lumino#44) were "truly" backwards compatible. In my view, the answer is yes, and the real problem stems from an unsafe "widening" of the `IRenderer.render` typing made on the Jupyterlab side in #7700. The good news there is that the changes from #7700 were only ever included in the beta releases of jlab 2.0.

Thus, I think the way forward is to pull in this PR and do a release. I'll add an issue with the full details of the typing snafu, and we can debate whether to rollback Lumino there (but like I said, I don't think we need to).

blink1073 commented:
Sounds good, onward!
